telemetry with batching, persistence, and improved logging

Niks988 opened the pull request:
# Description
- Add event batching (50 events, 5s timeout) for better performance
- Implement persistent queue (10K events) to prevent event loss during QuestDB downtime
- Add connection health monitoring and auto-reconnection
- Improve logging with clear batch and queue status
- Add graceful shutdown with event flushing
- Eliminate ECONNRESET spam with error rate limiting
- Add fallback mechanisms and memory leak protection

Niks988 committed:
telemetry with batching, persistence, and improved logging

Niks988 committed:
improved proof logging, other improvements

Niks988 committed:
improved proof logging, other improvements

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+    constructor() {
+        this.localSender = null;
+        this.lastErrorLogTime = 0;
+        this.errorLogInterval = 60000; // 1 minute between error logs
```
Move all this constants to `constants/constants.js`

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+        this.lastSuccessfulSend = Date.now();
+        this.maxTimeWithoutSuccess = 5 * 60 * 1000; // 5 minutes
+        
+        // Event persistence for when QuestDB is down
```
Should this be stored in memory, if nodes restarts, it's lost, and increases the memory footprint of the node?

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+        try {
+            this.localSender = Sender.fromConfig(this.config.localEndpoint);
+            this.retryAttempts = 0;
+            this.telemetryStats.connectionStatus = 'connected';
```
connected should be constants

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+            this.lastSuccessfulSend = Date.now();
+            this.logger.debug('QuestDB local sender created successfully');
+        } catch (error) {
+            this.telemetryStats.connectionStatus = 'failed';
```
failed should be constants

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+    startHealthMonitoring() {
+        this.healthCheckTimer = setInterval(() => {
+            this.logHealthStatus();
+        }, this.healthCheckInterval);
```
`healthCheckInterval` shouldn't be field it should be constant

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+    async recreateConnection() {
+        try {
+            if (this.localSender) {
+                await this.localSender.flush();
```
What does `flush` do?

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+                    
+                    table.symbol('operationId', event.operationId || 'NULL');
+                    table.symbol('blockchainId', event.blockchainId || 'NULL');
+                    table.symbol('name', event.name || 'NULL');
```
should this be 'NULL' as string, or it should just be null value?

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+            
+            // Log failed proof events (important to know)
+            if (proofEvents.length > 0) {
+                this.logger.error(`[TELEMETRY] Failed to send ${proofEvents.length} proof events: ${proofEvents.map(e => e.operationId).join(', ')}`);
```
If there is a lot of failed events this log will be huge

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+        
+        for (const event of events) {
+            try {
+                await this.retryWithBackoff(async () => {
```
Do we need to this sequantaly, can we rewrite this so it retries at same time for all of them by launching promises at same time and awaits for all of them

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+        // Calculate queue status
+        const queueStatus = this.persistentEventQueue.length > 0
+            ? `Queue: ${this.persistentEventQueue.length}/${this.maxPersistentQueueSize}`
+            : 'Queue: empty';
```
it shouldn't be empty but 0% let's have consistants logs

Mihajlo-Pavlovic commented on src/modules/telemetry/implementation/quest-telemetry.js:
```diff
@@ -1,12 +1,599 @@
+        }
+    }
+
+    logLifetimeProofSummary() {
```
Remove dead code

Mihajlo-Pavlovic commented:
Did partial review, can you address these comments, also address claued raised issues and liter errors

swamala committed:
fix: address PR review comments

   - Remove signal handler duplication to prevent shutdown race conditions
   - Fix telemetry cleanup method access path
   - Extract magic values to constants for better maintainability
   - Limit proof event logging to prevent huge logs
   - Refactor sequential retry to parallel for better performance
   - Fix queue status log consistency
   - Remove dead code (logLifetimeProofSummary)

   Addresses review comments from PR #4014

swamala committed:
Merge branch 'v8/develop' into improvements/telemetry-upgrades
